test/index.js

```diff
@@ -1,223 +1,420 @@
+                --i;
+                if (i < 0) {
+                    clearInterval(intervalHandle);
+                    return done(new Error('recluster did not gracefully handle the reload'));
+                }
+            }, tryToDoneEvery);
         });
-    })
-});
 
-runTest("server with an endless setInterval", function(t) {
-    lib.setServer('server-unclean.js', function(err) {
-        t.ok(!err, "should change to unending server");
-        lib.balancer.reload(function() {
-            setTimeout(lib.balancer.reload.bind(lib.balancer, function() {
-                setTimeout(checkDead, 400); // 400 = timeout + 100ms spawntime
-            }), 10);
-        });
-        function checkDead() {
-            t.equal(lib.balancer.workers.length, 2, "only 2 workers should be active");
-            t.end(); // this test will never end
-        }
-    })
-});
+        it('should properly hang up if exceeds timeout', function(done) {
+            //200ms request vs 100ms timeout = socket trashed
+            request.get('/200')
+                .end(function(err, res) {
+                    assert(reloaded, 'reloaded should be true');
+                    assert.equal(err.message, 'socket hang up');
+                });
 
-runTest("server with arguments", {file: 'server-with-args.js', args: ['fail']}, function(t) {
-    request({url: 'http://localhost:8000'}, function(err) {
-        t.equal(err, 404, 'should get 404 error');
-        t.end();
+            setImmediate(function() {
+                handler.reload();
+                handler.once('ready', function() {
+                    reloaded = true;
+                });
+            });
+
+            intervalHandle = setInterval(function() {
+                if (Object.keys(cluster.workers).length === 2) {
+                    clearInterval(intervalHandle);
+                    return done();
+                }
+                --i;
+                if (i < 0) {
+                    clearInterval(intervalHandle);
+                    return done(new Error('recluster did not gracefully handle the reload'));
+                }
+            }, tryToDoneEvery);
+        });
     });
-});
 
-/**
- * Termination tests
- * 1) dies ungracefully after 1/2 s (lib/server-die-halfsec.js)
- * 2) disconnects  IPC  after 1/2 s (lib/server-disconnect-halfsec.js)
- * 3) msgs 'disconnect' after 1/2 s (lib/server-msg-disconnect-halfsec.js)
- * Then test if
- * 1) they have been replaced quickly
- * 2) they have been killed after a while for (2) and (3)
- *
- * Recommended settings are:
- * respawn: 0.01 - for quick respawns
- * backoff: 0.01 - for quick respawns
- * workers: 2 - to make sure we're not testing only for one
- * timeout: 0.3 - to be able to test if timeout works
- * before the replacement tries to signal that its dead
- *
- * I know that timing-based tests are not perfect, but I have no better
- * idea at the moment.
- */
+    //@FIXME: help me here, no idea what we are supposed to test
```

Comment:
What the original example does esentially is switch to a server that refuses to gracefully shutdown when the IPC channel is closed. This is typical in e.g. scenarios where there is a DB connection pool that doesn't get terminated when the http server closes.

What the original test verifies is that after a cluster reload, the old workers will not respond to the new clients and that they're properly terminated after the timeout. This should be guaranteed by node's cluster module, but [there was an issue](https://github.com/doxout/recluster/issues/2) and I decided to write a test to try and reproduce it.